### PR TITLE
tests: benchmarks: timing_info: Add support for ARC, XTENSA, NIOS2, RISCV

### PR DESCRIPTION
**Pull request by AdithyaBaglody**
2 Parts to this PR. 

Part 1:
Added new KPIs for the userspace benchmarks.
-  Drop to user mode.
-  User thread Creation.
-  Syscall overhead.
-  Validation overhead k object init.
-  Validation overhead k object permission

Part 2: 
- Expand all possible KPIs to ARC, XTENSA, NIOS2 and RISCV32
- Provide userspace KPIs to only x86, ARM and ARC (the 3 architectures in which userspace is enabled)

### CODE REVIEW

File: arch/x86/core/userspace.S [hunk 1]
```diff
@@ -169,6 +169,19 @@ SECTION_FUNC(TEXT, _x86_userspace_enter)
 	 */
 	mov	%esp, %edi
 
+#ifdef CONFIG_EXECUTION_BENCHMARKING
```
Comment (andrewboie) [inline arch/x86/core/userspace.S]: You might be able to get away with moving this right before the "iret" since the code before it just formats the stack.

File: include/misc/timing_info.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#ifndef _TIMING_INFO__H_
+#define _TIMING_INFO__H_
+
+#ifdef CONFIG_EXECUTION_BENCHMARKING
+
+__syscall int k_dummy_syscall(void);
+__syscall u32_t userspace_read_timer_value(void);
+__syscall int validation_overhead_syscall(void);
+#endif  /* CONFIG_EXECUTION_BENCHMARKING */
+
```
Comment (andrewboie) [inline include/misc/timing_info.h]: You need an #include <syscalls/timing_info.h> at the bottom of this header.

File: include/misc/timing_info.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
+/*
```
Comment (andrewboie) [inline include/misc/timing_info.h]: The syscalls defined here are specific to a single test case.
This should *not* go in include/misc/, this needs to be in the same codebase as the test case.
We might need to extend the syscall auto-magic to have additional search paths,

File: tests/benchmarks/timing_info/src/timing_info.h [hunk 4]
```diff
@@ -16,23 +16,58 @@
...
+#ifdef CONFIG_NRF_RTC_TIMER
+
 /* To get current count of timer, first 1 need to be written into
  * Capture Register and Current Count will be copied into corresponding
  * current count register.
  */
-#define TIMING_INFO_PRE_READ() (NRF_TIMER2->TASKS_CAPTURE[0] = 1)
-#define TIMING_INFO_OS_GET_TIME()  (NRF_TIMER2->CC[0])
-#define TIMING_INFO_GET_TIMER_VALUE()	TIMING_INFO_GET_CURRENT_TIME()
+#define TIMING_INFO_PRE_READ()        (NRF_TIMER2->TASKS_CAPTURE[0] = 1)
```
Comment (andrewboie) [inline tests/benchmarks/timing_info/src/timing_info.h]: We really need some arch-independent APIs for getting highest-resolution timing information instead of ifdefing our test cases like this

File: tests/benchmarks/timing_info/src/timing_info.h [hunk 5]
```diff
@@ -16,23 +16,58 @@
 #define DECLARE_VAR(profile, name) \
 	u64_t total_##profile##_##name##_time;
 
-/* NRF RTC TIMER runs ar very slow rate (32KHz), So in order to measure
- * Kernel starts a dedicated timer to measure kernel stats.
+/* Stack size for all the threads created in this benchmark */
+#define STACK_SIZE 1024
```
Comment (andrewboie) [inline tests/benchmarks/timing_info/src/timing_info.h]: `+ CONFIG_TEST_EXTRA_STACKSIZE` (or whatever it is)

File: tests/benchmarks/timing_info/prj_userspace.conf [hunk 6]
```diff
@@ -0,0 +1,10 @@
+CONFIG_TEST=y
+CONFIG_EXECUTION_BENCHMARKING=y
+CONFIG_TEST_RANDOM_GENERATOR=y
+CONFIG_HEAP_MEM_POOL_SIZE=256
+CONFIG_MAIN_STACK_SIZE=2048
```
Comment (andrewboie) [inline tests/benchmarks/timing_info/prj_userspace.conf]: why is main stacksize increased?

File: tests/benchmarks/timing_info/prj_userspace.conf [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+CONFIG_FORCE_NO_ASSERT=y
+
+#Disable Userspace
```
Comment (andrewboie) [inline tests/benchmarks/timing_info/prj_userspace.conf]: comment is opposite of what you are doing here. this is redundant anyway since CONFIG_TEST selects CONFIG_TEST_USERSPACE and CONFIG_TEST_HW_STACK_PROTECTION

File: tests/benchmarks/timing_info/src/userspace_bench.c [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+	k_thread_access_grant(k_current_get(), &test_sema, NULL);
+
+
+	k_thread_create(&my_thread_user, my_stack_area, STACK_SIZE,
+			validation_overhead_user_thread,
+			NULL, NULL, NULL,
+			-1 /*priority*/, K_INHERIT_PERMS | K_USER, 0);
+
+
+	u32_t total_cycles_obj_init = (u32_t)
```
Comment (andrewboie) [inline tests/benchmarks/timing_info/src/userspace_bench.c]: I keep seeing lines like this, any way to macro it to make simpler?

File: tests/benchmarks/timing_info/testcase.yaml [hunk 9]
```diff
@@ -1,4 +1,9 @@
 tests:
-  benchmark.timing:
+  benchmark.timing.default_kernel:
     arch_whitelist: x86 arm
     tags: benchmark
+  benchmark.timing.userspace:
+    filter: CONFIG_ARCH_HAS_USERSPACE
+    extra_args: CONF_FILE=prj_userspace.conf
+    arch_whitelist: x86 arm
```
Comment (andrewboie) [inline tests/benchmarks/timing_info/testcase.yaml]: how hard would it be to add ARC support?
just need to add another ifdef block for the timer right?

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]

File: arch/x86/core/userspace.S (hunk 1 above)
Comment (AdithyaBaglody) [reply arch/x86/core/userspace.S]: Sure I can move it.

File: tests/benchmarks/timing_info/prj_userspace.conf (hunk 6 above)
Comment (AdithyaBaglody) [reply tests/benchmarks/timing_info/prj_userspace.conf]: actaully no specific reason. These configs are taken from the old prj.conf file.

File: tests/benchmarks/timing_info/testcase.yaml (hunk 9 above)
Comment (AdithyaBaglody) [reply tests/benchmarks/timing_info/testcase.yaml]: currently adding the support. i will be creating another PR once that is done.

File: tests/CMakeLists.txt [hunk 10]
```diff
@@ -3,3 +3,6 @@ add_subdirectory_if_kconfig(ztest)
 zephyr_include_directories_ifdef(CONFIG_TEST
   $ENV{ZEPHYR_BASE}/tests/include
   )
+
+zephyr_syscall_include_ifdef(CONFIG_EXECUTION_BENCHMARKING
+  "${CMAKE_CURRENT_SOURCE_DIR}/benchmarks/timing_info/src")
```
Comment (andrewboie) [inline tests/CMakeLists.txt]: why is this here?
this should go in the CMakeLists.txt for the specific timing_info test!

File: tests/benchmarks/timing_info/testcase.yaml [hunk 11]
```diff
@@ -1,4 +1,9 @@
 tests:
-  benchmark.timing:
-    arch_whitelist: x86 arm
+  benchmark.timing.default_kernel:
+    arch_whitelist: x86 arm arc
```
Comment (nashif) [inline tests/benchmarks/timing_info/testcase.yaml]: shouldn't this whitelist more architectures, actually should run all?

**Review by nashif — CHANGES_REQUESTED** [review]

File: tests/benchmarks/timing_info/testcase.yaml (hunk 11 above)
Comment (AdithyaBaglody) [reply tests/benchmarks/timing_info/testcase.yaml]: yeah the remaining architectures should come here. I missed it

File: tests/benchmarks/timing_info/testcase.yaml [hunk 12]
```diff
@@ -1,4 +1,9 @@
 tests:
-  benchmark.timing:
-    arch_whitelist: x86 arm
+  benchmark.timing.default_kernel:
+    arch_whitelist: x86 arm arc xtensa nios2 riscv32
```
[the remark this answers is not in the document]
Comment (AdithyaBaglody) [reply tests/benchmarks/timing_info/testcase.yaml]: yep makes sense. Done.

**Comment by nashif** [comment]
@AdithyaBaglody  please rebase. Did you test this on ARC? it fails for me using the nsim simulator:

```
***** Booting Zephyr OS v1.12.0-937-g9189109847 *****
starting test - Time Measurement
Timing Results: Clock Frequency: 10 MHz
Exception vector: 0x6, cause code: 0x2, parameter 0x4
Address 0xfffffffc
Current thread ID = 0x8000001c
Faulting instruction address = 0x29a4
Fatal fault in essential thread! Spinning...
```

**Comment by nashif** [comment]
@AdithyaBaglody great, thanks, I just verified it on nsim_sem and emsk

**Review by nashif — APPROVED** [review]
